pymongo/pool.py

```diff
@@ -1415,7 +1415,7 @@ def _get_socket(self, all_credentials):
                             self._pending -= 1
                             self._max_connecting_cond.notify()
             sock_info.check_auth(all_credentials)
-        except Exception:
+        except BaseException:
```

Comment:
This doesn't really make sense to me. Sure we now catch gevent.Timeout, but won't gevent again raise _another_ timeout when we try to acquire the lock below on line 1422? In that case we might as well have not caught the exception anyway.

Reply:
You know what, I was on the old 3.11 branch...

The latest master uses contexts instead for acquiring / releasing, which should prevent the lock from remaining acquired... Let me try reproducing on master, and if it's fixed on master, we can just use an RC build from there.

Reply:
(then we can just close this)

Reply:
So as followup, it still causes problems on 3.11.4 because the except block isn't executed. `BaseException`, or a bare `except:` still fix this, because we don't nest timeouts currently.

Considering a worst case scenario where you have infinite timeouts waiting to fire, you would have a timeout fire at each:
- close_socket
- with self.size_cond
-  self.size_cond.notify (not 100%, this may not block and therefore not be an eligible point to fire a timeout)
- event_listeners.publish_connection_check_out_failed, if the listener executes blocking code that would call the hub's switch() method

Given we only have 1 active at a time, this isn't an issue for us in particular, but would be if someone did use stacked timeouts and they all fired in a worst-case scenario.

Reply:
It seems like it might be necessary to subclass Pool as GeventPool, I see that custom Pool classes can be provided, so that's what I'll try next. Thanks for being responsive though :)